README.md
# VMware Guest OS Disk Space Monitor

Monitors VMware guest OS disk space and triggers an Opsgenie alert when the available free space falls below the defined threshold

## Requirements

- [Powershell](https://docs.microsoft.com/en-us/powershell/)
  - _Note: Developed and tested on Windows 10 with Powershell version 5.1.19041.1023, code is likely to work on PSVersion >= 3, but untested.  Based on my experience porting to PS Core on Linux is possible, but will likely require some code modifications._
- [VMware.PowerCLI](https://www.powershellgallery.com/packages/VMware.PowerCLI)
- [PsIni](https://www.powershellgallery.com/packages/PsIni/3.1.2)
- [PSWriteHTML](https://www.powershellgallery.com/packages/PSWriteHTML)
- [Posh-Syslog](https://www.powershellgallery.com/packages/Posh-SYSLOG)
- [Chocolatey](https://chocolatey.org/)
- [NSSM](https://nssm.cc/download)

## Installation

- Download code from GitHub
    - _Note:  If you don't have Git installed you can also just grab the zip:  https://github.com/CC-Digital-Innovation/vmware_guest_diskspace_mon/archive/master.zip_

    ```Powershell
    git clone https://github.com/CC-Digital-Innovation/vmware_guest_diskspace_mon.git
    ```

- Copy config.ini.example to config.ini
- Update settings in config.ini file.

## Usage

- Install Requirements from Powershell command prompt

    ```Powershell
    Start-Process Powershell -Verb RunAs
    Set-ExecutionPolicy -Confirm:$false Unrestricted -Force
    Get-ExecutionPolicy
    Install-Module -Confirm:$false -Name PsIni
    .\requirements.ps1
    ```

- From Powershell command prompt

    ```Powershell
    .\check_diskspace.ps1
    ```

- Install as Windows service using nssm_helper.ps1

    _Note: Be sure to update the NSSM sevicename and scriptpath in the config.ini_

    ```Powershell
    .\nssm_helper.ps1
    ```
    _Note:  You will need to update the service to run as a user with admin privledge._


- From Windows command prompt or from task scheduler

    _Note: Running as a service is preferred, if you want to run as a scheduled task, you will need to modify the code to remove the while loop. I may update this to happen automatically in a future release depending on a config.ini setting._
    ```Powershell
    start powershell "& "./check_diskspace.ps1"
    ```



## Compatibility

This is was built and tested on Windows 10 with Powershell Version 5.1.19041.1023, but most likely work on any Windows system with Powershell >= Version 2.

## Disclaimer

The code provided in this project is an open source example and should not be treated as an officially supported product. Use at your own risk. If you encounter any problems, please log an [issue](https://github.com/CC-Digital-Innovation/vmware_guest_diskspace_mon/issues).

## Contributing

1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request ツ

## History

- See [CHANGELOG.md](https://github.com/CC-Digital-Innovation/vmware_guest_diskspace_mon/blob/main/CHANGELOG.md)

## Credits

Rich Bocchinfuso <<rbocchinfuso@gmail.com>>

## License

MIT License

Copyright (c) [2021] [Richard J. Bocchinfuso]

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
<!-- markdownlint-disable MD001 MD009 MD012 MD024 MD032-->

# Changelog

## [Unreleased]

- TODO
  - Config parameters to enable | disable functions

## [0.6.0] - 2021-07-21

### Added

- sendReport function to generate and email reports
- napTime function to visually monitor cycle time


### Fixed

- Replaced hard coded path with $PSScriptRoot
- Email report logic


## [0.5.0] - 2021-07-19

### Added

- Logging to output.log for debugging
- Added logic for multiple vCenter connections


### Fixed

- Some issues with requirement.ps1


## [0.4.0] - 2021-07-01

### Added

- Cycle time and loop to prepare for service creation
- nssm-helper.ps1 to simplify service creation
- requirement.ps1 to install all requirements

## [0.3.0] - 2021-06-30

### Added

- Code to parse responders and tags from config.ini and place into array to properly pass Opsgenie body JSON payload
- Syslog function


### Fixed

- Parsing responders and tags and properly constructing the array


## [0.2.0] - 2021-06-29

### Added

- config.ini file for all configuration parameters


### Security

- Removed sensative config info from code and put into config.ini
- Added config.ini.example and updated .gitignore to exclude .ini from commits

## [0.1.0] - 2021-06-29

### Added

- Pulling and comparing VM guest OS disk space to threshold
- Opsgenie function to sent alert to Opsgenie when disk space falls below the defined threshold



nssm-helper.ps1
#requires -version 2
<#
.SYNOPSIS
  Helper script to create service
.DESCRIPTION
  Creates and starts windows service
.INPUTS
  ini file
.NOTES
  Version:        0.4.0
  Author:         Rich Bocchinfuso
  Creation Date:  06/29/2021
  Purpose/Change: Helper script to create windows servcie
#>


#---------------------------------------[Declarations]------------------------------------

# Script Version
$ScriptVersion = "0.0.4"


#---------------------------------------[Initialisations]----------------------------------

# Read config paramaters from config.ini file using PsIni
$CONFIG = Get-IniContent "config.ini"

# Environment
$nssm = (Get-Command nssm).Source
$powershell = (Get-Command powershell).Source
$arguments = '-ExecutionPolicy Bypass -NoProfile -File "{0}"' -f $scriptPath
$serviceName = $CONFIG["NSSM"]["servicename"]
$scriptPath = $CONFIG["NSSM"]["scriptpath"]

#---------------------------------------[Functions]---------------------------------------


#---------------------------------------[Execution]---------------------------------------

& $nssm install $serviceName $powershell $arguments
& $nssm status $serviceName
Start-Service $serviceName
Get-Service $serviceName



requirements.ps1
#requires -version 5
<#
.SYNOPSIS
  Script to install requirements
.DESCRIPTION
  Installs requirements
.INPUTS
  None
.NOTES
  Version:        0.6.0
  Author:         Rich Bocchinfuso
  Creation Date:  06/29/2021
  Purpose/Change: Initial script development to monitor and alert on VM guest volume free space
#>

# Start-Process Powershell -Verb RunAs
# Set-ExecutionPolicy -Confirm:$false Unrestricted -Force
# Get-ExecutionPolicy
# Install-Module -Confirm:$false -Name PsIni

#---------------------------------------[Initializations]---------------------------------

# Install PSIni Module
Install-Module -Confirm:$false -Name PsIni

# Read config paramaters from config.ini file using PsIni
$CONFIG = Get-IniContent "$PSScriptRoot\config.ini"

# Script Version
$RelaseVersion = $CONFIG["Environment"]["release"]


#---------------------------------------[Declarations]------------------------------------

$modules = $CONFIG["Powershell"]["modules"]

#---------------------------------------[Functions]---------------------------------------

function InstallModules {
  $ModuleArray = $modules.Split(",")
  $ModuleArray
  foreach ($psm in $ModuleArray) {
    if (Get-Module -ListAvailable -Name $psm) {
      Write-Host ($psm + " is already installed")
    }
    else {
      try {
        Install-Module -Name $psm -AllowClobber -Confirm:$False -Force
      }
      catch [Exception] {
        $_.message
        exit
      }
    }
  }
}


#---------------------------------------[Execution]---------------------------------------


# Check to see if you are running as Administator, if not elevate privledge
if (-NOT ([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole] "Administrator")) {  
  $arguments = "& '" + $myinvocation.mycommand.definition + "'"
  Start-Process powershell -Verb runAs -ArgumentList $arguments
  Break
}


Set-PSRepository -Name 'PSGallery' -InstallationPolicy Trusted

InstallModules

# Install Chocolaty and NSSM
Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; Invoke-Expression ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))
& choco install nssm --confirm

Set-PowerCLIConfiguration -Confirm:$false -ParticipateInCEIP:$false
Set-PowerCLIConfiguration -Confirm:$false -InvalidCertificateAction Ignore



nssm_helper.ps1
#requires -version 5
<#
.SYNOPSIS
  Helper script to create Windows service
.DESCRIPTION
  Creates and starts Windows service
.INPUTS
  ini file
.NOTES
  Version:        0.5.0
  Author:         Rich Bocchinfuso
  Creation Date:  06/29/2021
  Purpose/Change: Helper script to create windows servcie
#>


#---------------------------------------[Initializations]---------------------------------


# Read config paramaters from config.ini file using PsIni
$CONFIG = Get-IniContent "$PSScriptRoot\config.ini"

# Script Version
$RelaseVersion = $CONFIG["Environment"]["release"]


#---------------------------------------[Declarations]------------------------------------

# Environment
$nssm = (Get-Command nssm).Source
$powershell = (Get-Command powershell).Source
$arguments = '-ExecutionPolicy Bypass -NoProfile -File "{0}"' -f $scriptPath
$serviceName = $CONFIG["NSSM"]["servicename"]
$scriptPath = $CONFIG["NSSM"]["scriptpath"]

#---------------------------------------[Functions]---------------------------------------


#---------------------------------------[Execution]---------------------------------------

& $nssm install $serviceName $powershell $arguments
& $nssm status $serviceName
Start-Service $serviceName
Get-Service $serviceName

